ocaml/libs/tracing/tracing.ml

```diff
@@ -92,11 +94,85 @@ module Spans = struct
               Hashtbl.replace spans key (span :: span_list)
     )
 
+  let mark_finished ~(span : Span.t) =
+    Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+        let key = span.context.trace_id in
+        match Hashtbl.find_opt spans key with
+        | None ->
+            debug "Span does not exist or already finished"
+        | Some span_list -> (
+            List.filter
+              (fun x -> x.Span.context <> span.context)
+              span_list
+            |> Hashtbl.replace spans key ;
+            match Hashtbl.find_opt finished_spans key with
+            | None ->
+                Hashtbl.add finished_spans key [span]
+            | Some span_list ->
+                if List.length span_list < 1000 then
+                  Hashtbl.replace finished_spans key (span :: span_list)
+          )
+    )
+
+  let assert_finished x =
+    match x with
+    | None ->
+        false
+    | Some (span : Span.t) -> (
+      match Hashtbl.find_opt finished_spans span.context.trace_id with
+      | None ->
+          false
+      | Some span_list ->
+          List.exists (fun x -> x = span) span_list
+    )
+
   let since () =
     Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
-        let copy = Hashtbl.copy spans in
-        Hashtbl.clear spans ; copy
+        let copy = Hashtbl.copy finished_spans in
+        Hashtbl.clear finished_spans ;
+        copy
     )
+
+  module GC = struct
+    let span_timeout = ref 86400.
+
+    let span_timeout_thread = ref None
+
+    let gc_inactive_spans () =
+      Hashtbl.iter
```

Comment:
This needs a mutex: it modifies a global data structure and that is not safe to do concurrently.
See also https://v2.ocaml.org/api/Hashtbl.html 
"Alert unsynchronized_access. Unsynchronized accesses to hash tables are a programming error." the hashtable doesn't have any internal locking.